src/metabase/db/custom_migrations.clj

```diff
@@ -352,3 +361,88 @@
   (run! update-card-row-on-downgrade-for-dashboard-tab
         (eduction (map :dashboard_id) (t2/reducible-query {:select-distinct [:dashboard_id]
                                                            :from            [:dashboard_tab]}))))
+
+(define-migration RevisionMigrateLegacyColumnSettingsFieldRefs
+  (let [update-one! (fn [{:keys [id object]}]
+                      (let [object  (json/parse-string object)
+                            updated (update object "visualization_settings" update-legacy-field-refs-in-viz-settings)]
+                        (when (not= updated object)
+                          (t2/query-one {:update :revision
+                                         :set    {:object (json/generate-string updated)}
+                                         :where  [:= :id id]}))))]
+    (run! update-one! (t2/reducible-query {:select [:id :object]
+                                           :from   [:revision]
+                                           :where  [:and
+                                                    [:= :model "Card"]
+                                                    [:or
+                                                     ;; these match legacy field refs in column_settings
+                                                     [:like :object "%ref\\\\\",[\\\\\"field-id%"]
+                                                     [:like :object "%ref\\\\\",[\\\\\"field-literal%"]
+                                                     [:like :object "%ref\\\\\",[\\\\\"fk->%"]
+                                                     ;; MySQL with NO_BACKSLASH_ESCAPES disabled:
+                                                     [:like :object "%ref\\\\\\\",[\\\\\\\"field-id%"]
+                                                     [:like :object "%ref\\\\\\\",[\\\\\\\"field-literal%"]
+                                                     [:like :object "%ref\\\\\\\",[\\\\\\\"fk->%"]]]}))))
+
+(define-reversible-migration RevisionAddJoinAliasToColumnSettingsFieldRefs
```

Comment:
Explanation of how this migration works:

For each alias in `(get-in card ["dataset_query" "query" "joins"])`, we add new entries in `column_settings`.

For example:

```
If this was a entry in column_settings before the migration:
["ref" ["field" id nil]] {:column_title "A"}

And the query had 2 aliases in the query:
"Joined table 1", "Joined table 2"

After migration, there would be 3 entries in the column_settings table for each entry before the migration:
["ref" ["field" id nil]] {:column_title "A"}
["ref" ["field" id {"join-alias" "Joined table 1"}]] {:column_title "A"}
["ref" ["field" id {"join-alias" "Joined table 2"}]] {:column_title "A"}
```

Even if `["field" id {"join-alias" "Joined table 1"}]]` and `["field" id {"join-alias" "Joined table 2"}]]` aren't actually field refs in the card, this method preserves the observed column settings from the perspective of the user. At least one of them will be a real field ref of the card. The ones that aren't don't matter.